pkg/controllers/mcs/mcs_controller.go

```diff
@@ -0,0 +1,398 @@
+				workMeta.GetNamespace(), workMeta.GetName(), clusterName, err))
+			errs = append(errs, err)
+		}
+	}
+	if len(errs) != 0 {
+		return syncClusters, errors.NewAggregate(errs)
+	}
+
+	return syncClusters, nil
+}
+
+func (c *MCSController) updateMCSStatus(
+	ctx context.Context,
+	mcs *networkingv1alpha1.MultiClusterService,
+) error {
+	currentServiceAppliedCondition := meta.FindStatusCondition(mcs.Status.Conditions, MCSServiceAppliedConditionType)
+	if currentServiceAppliedCondition == nil {
+		mcs.Status.Conditions = append(mcs.Status.Conditions, metav1.Condition{
+			Message:            "Service is propagated to target clusters.",
+			Reason:             "ServiceAppliedSuccess",
+			Status:             metav1.ConditionTrue,
+			Type:               MCSServiceAppliedConditionType,
+			LastTransitionTime: metav1.Now(),
+		})
+	}
+
+	return retry.RetryOnConflict(retry.DefaultRetry, func() (err error) {
+		updateErr := c.Status().Update(ctx, mcs)
+		if updateErr == nil {
+			return nil
+		}
+
+		updated := &networkingv1alpha1.MultiClusterService{}
+		if err = c.Get(ctx, client.ObjectKey{Namespace: mcs.Namespace, Name: mcs.Name}, updated); err == nil {
+			updated.Status.Conditions = mcs.Status.Conditions
+			mcs = updated
+		} else {
+			klog.Errorf("Failed to get updated MultiClusterService(%s/%s): %v", mcs.Namespace, mcs.Name, err)
+		}
+		return updateErr
+	})
+}
+
+func (c *MCSController) getMultiClusterServiceID(mcs *networkingv1alpha1.MultiClusterService) (string, error) {
+	id := util.GetLabelValue(mcs.GetLabels(), networkingv1alpha1.MultiClusterServicePermanentIDLabel)
+	if id == "" {
+		id = uuid.New().String()
+		mcs.Labels = util.DedupeAndMergeLabels(mcs.Labels, map[string]string{networkingv1alpha1.MultiClusterServicePermanentIDLabel: id})
+		if err := c.Client.Update(context.TODO(), mcs); err != nil {
+			return id, err
+		}
+	}
+
+	return id, nil
+}
+
+// MCSContainCrossClusterType checks weather the MultiClusterService contains CrossCluster type.
+func MCSContainCrossClusterType(mcs *networkingv1alpha1.MultiClusterService) bool {
+	for _, t := range mcs.Spec.Types {
+		if t == networkingv1alpha1.ExposureTypeCrossCluster {
+			return true
+		}
+	}
+	return false
+}
+
+// SetupWithManager creates a controller and register to controller manager.
+func (c *MCSController) SetupWithManager(mgr controllerruntime.Manager) error {
+	mcsPredicateFunc := predicate.Funcs{
+		CreateFunc: func(e event.CreateEvent) bool {
+			mcs := e.Object.(*networkingv1alpha1.MultiClusterService)
+			return MCSContainCrossClusterType(mcs)
+		},
+		UpdateFunc: func(e event.UpdateEvent) bool {
+			mcsOld := e.ObjectOld.(*networkingv1alpha1.MultiClusterService)
+			mcsNew := e.ObjectNew.(*networkingv1alpha1.MultiClusterService)
+			if !MCSContainCrossClusterType(mcsOld) && !MCSContainCrossClusterType(mcsNew) {
+				return false
+			}
+
+			// We only care about the update events below:
+			if equality.Semantic.DeepEqual(mcsOld.Annotations, mcsNew.Annotations) &&
```

Comment:
Why is this judgment necessary?

Reply:
Currently, there don't need to pay attention to other changes, such as label.